File: Pool.lua
--[[--------------------------------------------------------------------
    Copyright (C) 2013, 2014 Johnny C. Lam.
    See the file LICENSE.txt for copying permission.
--]]--------------------------------------------------------------------

-- Simple resource pool.
local OVALE, Ovale = ...
local OvalePool = {}
Ovale.OvalePool = OvalePool

--<private-static-properties>
-- Profiling set-up.
local OvaleProfiler = Ovale.OvaleProfiler

local assert = assert
local setmetatable = setmetatable
local tinsert = table.insert
local tostring = tostring
local tremove = table.remove
local wipe = wipe

-- Register for profiling.
OvaleProfiler:RegisterProfiling(OvalePool, "OvalePool")
--</private-static-properties>

--<public-static-properties>
OvalePool.name = "OvalePool"
OvalePool.pool = nil
OvalePool.size = 0
OvalePool.unused = 0
OvalePool.__index = OvalePool
--</public-static-properties>

--<public-static-methods>
do
	-- Class constructor
	setmetatable(OvalePool, { __call = function(self, ...) return self:NewPool(...) end })
end

function OvalePool:NewPool(name)
	name = name or self.name
	local obj = setmetatable({ name = name }, self)
	obj:Drain()
	return obj
end

function OvalePool:Get()
	OvalePool:StartProfiling(self.name)
	assert(self.pool)
	local item = tremove(self.pool)
	if item then
		self.unused = self.unused - 1
	else
		self.size = self.size + 1
		item = {}
	end
	OvalePool:StopProfiling(self.name)
	return item
end

function OvalePool:Release(item)
	OvalePool:StartProfiling(self.name)
	assert(self.pool)
	self:Clean(item)
	wipe(item)
	tinsert(self.pool, item)
	self.unused = self.unused + 1
	OvalePool:StopProfiling(self.name)
end

function OvalePool:Clean(item)
	-- virtual function; override as needed.
end

function OvalePool:Drain()
	OvalePool:StartProfiling(self.name)
	self.pool = {}
	self.size = self.size - self.unused
	self.unused = 0
	OvalePool:StopProfiling(self.name)
end

function OvalePool:DebuggingInfo()
	Ovale:Print("Pool %s has size %d with %d item(s).", tostring(self.name), self.size, self.unused)
end
--</public-static-methods>

File: src/tools/Pool.ts
import { assert, tostring, wipe, LuaArray } from "@wowts/lua";
import { insert, remove } from "@wowts/table";
import { format } from "@wowts/string";

export class OvalePool<T> {
    pool: LuaArray<T> = {};
    size = 0;
    unused = 0;
    name: string;

    constructor(name: string) {
        this.name = name || "OvalePool";
    }

    get() {
        // OvalePool.StartProfiling(this.name);
        assert(this.pool);
        let item = remove(this.pool);
        if (item) {
            this.unused = this.unused - 1;
        } else {
            this.size = this.size + 1;
            item = <T>{};
        }
        // OvalePool.StopProfiling(this.name);
        return item;
    }
    release(item: T): void {
        // OvalePool.StartProfiling(this.name);
        assert(this.pool);
        this.clean(item);
        insert(this.pool, item);
        this.unused = this.unused + 1;
        // OvalePool.StopProfiling(this.name);
    }
    drain(): void {
        //OvalePool.StartProfiling(this.name);
        this.pool = {};
        this.size = this.size - this.unused;
        this.unused = 0;
        //OvalePool.StopProfiling(this.name);
    }
    debuggingInfo(): string {
        return format(
            "Pool %s has size %d with %d item(s).",
            tostring(this.name),
            this.size,
            this.unused
        );
    }

    clean(item: T): void {
        wipe(item);
    }
}
